_posts/2018-05-20-Y-The-Most-Beautiful-Idea-in-Computer-Science.md

```diff
@@ -0,0 +1,717 @@
+---
+layout: post
+title: "Y: The Most Beautiful Idea in Computer Science explained in JavaScript"
+author: Lucas Fernandes da Costa
+tags: functional-programming combinators javascript y-combinator
+---
+
+In this post, we will talk about **one of the most beautiful ideas in computer science: the Y-Combinator**. And no, I'm not talking about [the VC firm in Silicon Valley](http://www.ycombinator.com/), even though this post will explain why they've got this name.
+
+Put in simple terms, **the Y-Combinator (also known as the fixed-point combinator) is a way of doing recursion in a language that does not explicitly supports it**.
+
+Let's say you want to implement a recursive [factorial function](https://en.wikipedia.org/wiki/Factorial). In JavaScript, for example, you could simply do this:
+
+```js
+const factorial = n => n === 0 ? 1 : n * factorial(n - 1);
+```
+
+But what if you couldn't use any "names"? What if JavaScript didn't allow you to call `factorial` inside of `factorial`, how would you implement it?
+
+In that case, you'd use the Y-Combinator.
+
+I guarantee you **this will be one of the most amazing journeys you will go through in your life as a programmer**. The first time deriving the Y-Combinator is an indescribable joy, so get comfortable in your chair, bed, or whatever and let's get started.
+
+I recommend that you **follow this post with a notebook or on Node's REPL**, that will help you understand all the steps we'll go through and test things by yourself. **Don't be afraid of trying things out and testing things, doing your own exploration plays a great part in understanding how Y works**.
+
+Also, if you want to learn more about recursion [I've got a very popular post about it in this very same blog](/2017/05/08/All-About-Recursion-PTC-TCO-and-STC-in-JavaScript.html). It's a bit old but it remains amazing and very informative.
+
+Before we start I also need to warn you that **deriving the Y-Combinator for the first time is hard** and that **it's normal to struggle with it**. So don't feel bad if you have to read this blog post more than once in order to get it. **I tried to make it very detailed and show every single step of what is happening on each example but [if you feel like you're advanced enough and want to read the short version of it, just click here**](/2018/05/20/Y-The-Most-Beautiful-Idea-in-Computer-Science.html#finding-the-y-combinator).
```

Comment:
You can do relative links in Markdown, `[link text](#finding-the-y-combinator]`

Reply:
Amazing, I didn't know that!

Will fix. Thanks 👍